File: moeali/reversedLinkedList.py
# Given the head of a singly linked list, reverse the list, and return the reversed list.

class Solution(object):
    def reverseList(self, head):
        prev_node = None
        curr_node = head
        while curr_node is not None:
            next_node = curr_node.next
            curr_node.next = prev_node
            prev_node = curr_node
            curr_node = next_node
        return prev_node

File: gabe/reversed_linked_list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reversal(ListNode* prev, ListNode* head) {
        if(head->next == nullptr) {
            head->next = prev;
            return head;
        }

        ListNode * nextNode = head->next;
        head->next = prev;
        return reversal(head, nextNode);
    }

    ListNode* reverseList(ListNode* head) {
        if (head != nullptr) {
            return reversal(nullptr, head);
        }
        return head;
    }

};